### ISSUE TITLE
Consumer Offset Re-Positioning (Replay)

### ISSUE THREAD

**Issue by travis-minke-sap** [knative-sandbox/eventing-kafka#477]
**Problem**
The short-term persistence of events in the KafkaChannel is a key differentiator from other Channel implementations. This persistence provides the ability for supporting advanced use cases, such as being able to replay prior events. The ability to replay past events will allow subscribers who have experienced downtime to receive events they might have otherwise missed.

This issue covers adding the ability to re-position (backwards / forwards) the ConsumerGroup offset(s) associated with a specific Subscription. This will be done by date instead of by exposing the specific offset(s) for each partition.

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)**
Event consumer

**Exit Criteria**
- Repositioning will be based on a Date/Timestamp to start with. Once this is working we can evaluate the need for specifying offsets directly, but that is more complicated as each Partition/Consumer in the ConsumerGroup has a unique offset, and we would have to expose this fine-grained mapping of offsets to partition.
- All ConsumerGroup instances associated with a Subscription will be repositioned to the appropriate offset corresponding to the user specified Date/Timestamp. These offsets will differ from each other but will all be tied to the same timestamp.
- A history of repositions should be stored as a custom annotation in the
  associated Subscription with content something like the following...
    ```yaml
    annotations:
      messaging.knative.dev/position-history: |
      {
        { "timestamp": "2021-01-10T08:14:00Z", "position": "2021-01-07T23:59:59Z" }
        { "timestamp": "2021-02-11T08:14:00Z", "position": "2021-02-08T23:59:59Z" }
        { "timestamp": "2021-03-12T08:14:00Z", "position": "2021-03-09T23:59:59Z" }
      }
    ```
- Controllers (distributed & consolidated) expose the ability for user to specify their intent to reposition a specific Subscription to a distinct timestamp. **!!! SEE CONTROLLER API SECTION BELOW !!!**
- The Controller will communicate, via the [control-protocol](https://github.com/knative-sandbox/control-protocol) or a similar mechanism, with the Dispatcher replicas to implement the offset change.

**Controller API**

It is unclear what form the Controller's API for this new feature should take.  The ability to dynamically change the offsets is a one-time ephemeral "action" that doesn't fit neatly into the standard K8S / Knative declarative approach.  How to best expose such things is being discussed in the Knative Eventing [Discussion #5136](https://github.com/knative/eventing/discussions/5136) and this issue is blocked until that is resolved.

**Developer Notes**

- Sarama provides the ability to get offsets for a specific topic / partition / timestamp.
  ```
  Client.GetOffset(topic string, partitionID int32, time int64) (int64, error)
  ```
- Sarama provides the ability to reset the offsets of a partition.
  ```
  ConsumerGroupSession.ResetOffset(topic string, partition int32, offset int64, metadata string)
  ```

**Time Estimate (optional):**
Large

**Additional context (optional)**
- Might be related to pause/resume capability.
- Control-protocol is very new, might not be ready for this?

**Comment by slinkydeveloper** [comment]
@travis-minke-sap I was wondering, how does it sounds to you to go ahead and **try** to implement the pause/resume capability in `Subscription` first? The reason I say this is:

* pause/resume already has an FT, a prototype (https://github.com/slinkydeveloper/control-data-plane-communication), the tools to implement it (https://github.com/knative-sandbox/control-protocol)
* Because the replay feature will add a non trivial amount of code, I think we should focus on pause/resume, which I believe it's easier to implement. I believe that implementing pause/resume first will make room for the code to implement the replay, but I may be wrong here, I can't predict the future :smile:  

In parallel, we can find work on finding a solution for the API issue highlighted here https://github.com/knative/eventing/discussions/5136 and after the pause/resume is ready, implementing the replay will be as simple as adding an additional control message handler!

Of course, we can work together on this, I can bootstrap the required code in `Subscription` for that.

**Comment by travis-minke-sap** [comment]
> I was wondering, how does it sounds to you to go ahead and try to implement the pause/resume capability in `Subscription` first?

I am ok with the approach, and we are definitely also interested in the Pause/Resume capability.  My only concern is the timeframe, since the Replay feature is a requirement I need to fulfill (internally), whereas Pause/Resume is just a nice-to-have.  I'd be happy to pair on the two of them as you suggest, but would need to have some reasonable idea that we could complete both of them in the next 2-3 months (mid June?).  You know more than I do about what is left for Pause/Resume, but I would think if we both have the time to dedicate to it that should be possible?  I don't know what other commitments are on your plate though?

**Comment by slinkydeveloper** [comment]
> My only concern is the timeframe, since the Replay feature is a requirement I need to fulfill (internally), whereas Pause/Resume is just a nice-to-have. I'd be happy to pair on the two of them as you suggest, but would need to have some reasonable idea that we could complete both of them in the next 2-3 months (mid June?)

I don't feel comfortable telling you "in 1/2 months this feature is ready", even worse given the fact that for Replay we still don't have the full picture of how it should work... 

I can tell you that subscription pause/resume is something it shouldn't take that long, I can put together a plan in some upstream issues, based on the findings I made in the prototype. But for the replay, before start talking about timeframes, we need to solve https://github.com/knative/eventing/discussions/5136 first. Without that, I can't picture in my mind how much it takes and what has to be done in order to implement it.

**Comment by travis-minke-sap** [comment]
Yeah, fair enough - and of course you're right we need to settle  [knative/eventing#5136](https://github.com/knative/eventing/discussions/5136) first - I was just looking for a quick gut-check of... _"no, I have 10 things to do before that"_,   or _"pause/resume won't be ready for 4 months at least"_ or something like that.

For the moment we can keep pursuing the design (# 5136) and then when we're settled on desired approach determine if it looks feasible in that timeframe, and if not what other options I might have ; )

**Comment by pierDipi** [comment]
I think we should not restrict this feature to be only usable by a Subscription but general for both Trigger and Subscription.

It seems worth evaluating what Kafka itself offers with the `/bin/kafka-consumer-groups.sh` script [1] regarding resetting offsets, which could be used as a reference to build the API on or even for implementing it.

A possible API might be:
```yaml
apiVersion: kafka.eventing.knative.dev/v1alpha1
kind: ResetOffsets # The CLI use `--reset-offsets`
metadata:
  # ...
spec:
  # Ref allows getting auth context, consumer group, and topic.
  ref: <KReference to Subscription or Trigger or whatever comes next>
  to: 
    # One of:
    datetime: <YYYY-MM-DDTHH:mm:SS.sss> # The CLI use `--to-datetime <YYYY-MM-DDTHH:mm:SS.sss>`
    current: # (Optional) - The CLI use `--to-current` 
    latest: # (Optional) - The CLI use `--to-latest`
    earliest: # (Optional) - The CLI use `--to-earliest`
    # Others ...
status:
  # ...
```

The channel reconciler will reconcile only `ResetOffsets` that reference a `Subscription`, the broker one will reconcile only `ResetOffsets` that reference a `Trigger`

___

Notes:
1. > ```
    > --reset-offsets                         Reset offsets of consumer group.       
    >                                         Supports one consumer group at the   
    >                                         time, and instances should be        
    >                                         inactive 
    > ```
    I think we need the pause/resume feature as a prerequisite for this feature since we can't reset offsets of a running consumer.
2. We should consider using the CLI and run it in a k8s Job to avoid reinventing the wheel.

___ 
[1]
```
Option                                  Description                            
------                                  -----------                            
--all-groups                            Apply to all consumer groups.          
--all-topics                            Consider all topics assigned to a      
                                          group in the `reset-offsets` process.
--bootstrap-server <String: server to   REQUIRED: The server(s) to connect to. 
  connect to>                                                                  
--by-duration <String: duration>        Reset offsets to offset by duration    
                                          from current timestamp. Format:      
                                          'PnDTnHnMnS'                         
--command-config <String: command       Property file containing configs to be 
  config property file>                   passed to Admin Client and Consumer. 
--delete                                Pass in groups to delete topic         
                                          partition offsets and ownership      
                                          information over the entire consumer 
                                          group. For instance --group g1 --    
                                          group g2                             
--delete-offsets                        Delete offsets of consumer group.      
                                          Supports one consumer group at the   
                                          time, and multiple topics.           
--describe                              Describe consumer group and list       
                                          offset lag (number of messages not   
                                          yet processed) related to given      
                                          group.                               
--dry-run                               Only show results without executing    
                                          changes on Consumer Groups.          
                                          Supported operations: reset-offsets. 
--execute                               Execute operation. Supported           
                                          operations: reset-offsets.           
--export                                Export operation execution to a CSV    
                                          file. Supported operations: reset-   
                                          offsets.                             
--from-file <String: path to CSV file>  Reset offsets to values defined in CSV 
                                          file.                                
--group <String: consumer group>        The consumer group we wish to act on.  
--help                                  Print usage information.               
--list                                  List all consumer groups.              
--members                               Describe members of the group. This    
                                          option may be used with '--describe' 
                                          and '--bootstrap-server' options     
                                          only.                                
                                        Example: --bootstrap-server localhost: 
                                          9092 --describe --group group1 --    
                                          members                              
--offsets                               Describe the group and list all topic  
                                          partitions in the group along with   
                                          their offset lag. This is the        
                                          default sub-action of and may be     
                                          used with '--describe' and '--       
                                          bootstrap-server' options only.      
                                        Example: --bootstrap-server localhost: 
                                          9092 --describe --group group1 --    
                                          offsets                              
--reset-offsets                         Reset offsets of consumer group.       
                                          Supports one consumer group at the   
                                          time, and instances should be        
                                          inactive                             
                                        Has 2 execution options: --dry-run     
                                          (the default) to plan which offsets  
                                          to reset, and --execute to update    
                                          the offsets. Additionally, the --    
                                          export option is used to export the  
                                          results to a CSV format.             
                                        You must choose one of the following   
                                          reset specifications: --to-datetime, 
                                          --by-period, --to-earliest, --to-    
                                          latest, --shift-by, --from-file, --  
                                          to-current.                          
                                        To define the scope use --all-topics   
                                          or --topic. One scope must be        
                                          specified unless you use '--from-    
                                          file'.                               
--shift-by <Long: number-of-offsets>    Reset offsets shifting current offset  
                                          by 'n', where 'n' can be positive or 
                                          negative.                            
--state [String]                        When specified with '--describe',      
                                          includes the state of the group.     
                                        Example: --bootstrap-server localhost: 
                                          9092 --describe --group group1 --    
                                          state                                
                                        When specified with '--list', it       
                                          displays the state of all groups. It 
                                          can also be used to list groups with 
                                          specific states.                     
                                        Example: --bootstrap-server localhost: 
                                          9092 --list --state stable,empty     
                                        This option may be used with '--       
                                          describe', '--list' and '--bootstrap-
                                          server' options only.                
--timeout <Long: timeout (ms)>          The timeout that can be set for some   
                                          use cases. For example, it can be    
                                          used when describing the group to    
                                          specify the maximum amount of time   
                                          in milliseconds to wait before the   
                                          group stabilizes (when the group is  
                                          just created, or is going through    
                                          some changes). (default: 5000)       
--to-current                            Reset offsets to current offset.       
--to-datetime <String: datetime>        Reset offsets to offset from datetime. 
                                          Format: 'YYYY-MM-DDTHH:mm:SS.sss'    
--to-earliest                           Reset offsets to earliest offset.      
--to-latest                             Reset offsets to latest offset.        
--to-offset <Long: offset>              Reset offsets to a specific offset.    
--topic <String: topic>                 The topic whose consumer group         
                                          information should be deleted or     
                                          topic whose should be included in    
                                          the reset offset process. In `reset- 
                                          offsets` case, partitions can be     
                                          specified using this format: `topic1:
                                          0,1,2`, where 0,1,2 are the          
                                          partition to be included in the      
                                          process. Reset-offsets also supports 
                                          multiple topic inputs.               
--verbose                               Provide additional information, if     
                                          any, when describing the group. This 
                                          option may be used with '--          
                                          offsets'/'--members'/'--state' and   
                                          '--bootstrap-server' options only.   
                                        Example: --bootstrap-server localhost: 
                                          9092 --describe --group group1 --    
                                          members --verbose                    
--version                               Display Kafka version. 
```

**Comment by travis-minke-sap** [comment]
Hi Pier - Thanks for the feedback!

My thoughts...
- I think the idea of making the Command CRD more generic (not specific to subscriptions but also triggers) makes sense.
- I don't think we have to have Pause/Resume in order to reposition the offset.   The ConsumerGroup will just be interrupted from processing wherever it is in the topic.  The Dispatcher logic can stop the ConsumerGroup if the Sarama library doesn't handle that automatically during the offset reposition.
- The idea of expanding this to the full-blown Kafka ConsumerGroups API really changes the scope of things.  I do see some rationale for exposing all this, but the intent was not to provide users with complete low-level control over the individual ConsumerGroups.  They can theoretically use the Kafka cmd line to do this today, and don't need to funnel that through a Knative CRD.  Instead I'm trying to provide select functionality (collectively rolling back all ConsumerGroups in a Subscription to a specific timestamp) in a simpler API that is more K8S/Knative.

I think what you're proposing could be another capability that is added separately if there is demand for a K8S pass-through to the Kafka cmd-line tools, but is different from the goal of this Issue.  Thoughts?

Thanks again!

**Comment by travis-minke-sap** [comment]
@pierDipi - I wanted to say thanks for providing that suggestion above - it sparked a conversation internally as to whether the Controller watching the new "Command" CRDs could just alter the offsets directly instead of having to communicate with the Dispatcher Pods.  I have some research to do in that regard, but assuming it is feasible then we are able to eliminate the need for the control-protocol, which is still POC level and might not be ready in time for this work.  I'm interested in everyone's opinion as to the acceptability of that versus what we'd been discussing related to the control-protocol?    Thanks again everyone - closing in on it ; )

**Comment by pierDipi** [comment]
> We we're trying to keep the Command CRD a bit more generic (maybe... KafkaConsumerCommand) so that it could contain other "commands" in the future - for now would just support "RepositionOffset".

What I'm trying to suggest is to look at what is out there and try to use the same language/API, and not to support all the stuff.

For example, why do we call it `RepositionOffset` when the command-line client, the Java API, and the Sarama API call it `ResetOffset`?

Also, If we create a CRD that supports multiple commands, I see an increased complexity in:
- granting access to individual commands rather than all of them.
- API or command versioning

> The idea of expanding this to the full-blown Kafka ConsumerGroups API really changes the scope of things. I do see some rationale for exposing all this, but the intent was not to provide users with complete low-level control over the individual ConsumerGroups. They can theoretically use the Kafka cmd line to do this today, and don't need to funnel that through a Knative CRD. Instead I'm trying to provide select functionality (collectively rolling back all ConsumerGroups in a Subscription to a specific timestamp) in a simpler API that is more K8S/Knative.

I agree!

> I have some research to do in that regard, but assuming it is feasible then we are able to eliminate the need for the control-protocol, which is still POC level and might not be ready in time for this work. I'm interested in everyone's opinion as to the acceptability of that versus what we'd been discussing related to the control-protocol? Thanks again everyone - closing in on it ; )

I haven't followed any conversation, so :man_shrugging:, but this was my assumption, it's a control plane only action.

I'm reading KIP-122:

> (1) It will get assignments from `describeGroup` operation
>
> (2) Then use `getAssignmentsToReset` to filter the assignments with the values defined by `–topic` of or `--all-topics`.
> 
> (3) Then get new offset by assignments using `prepareAssignmentsToReset`.
>
> (4) It will be only executed when the ConsumerGroup selected is inactive to avoid race conditions. 
>
> (5) It will be executed only if it is asked explictly. This will consist in create a Consumer using the same `group.id` and use:
> ```
> consumer.assign(List(topicPartition).asJava)
> consumer.seek(topicPartition, offset)
> consumer.commitSync()
> ```
> To change the offsets.

The race condition is referring to I guess is related to the fact that they need to support the `--shift-by n` option, which is out of scope for us.

The "is active" check is client-side https://github.com/apache/kafka/blob/617ee003223873da2c1c7f383d3416684c2d84c8/core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala#L451, so I guess we can do the reset directly to a specific timestamp using Sarama on the control plane without pausing consumers.

**Comment by slinkydeveloper** [comment]
> I have some research to do in that regard, but assuming it is feasible, then we are able to eliminate the need for the control-protocol, which is still POC level and might not be ready in time for this work

If it never gets used, it never becomes ready :) I think from now on we should invest on control-protocol for new features, in particular for this one where it fits very well

**Comment by travis-minke-sap** [comment]
> What I'm trying to suggest is to look at what is out there and try to use the same language/API, and not to support all the stuff.
> 
> For example, why do we call it `RepositionOffset` when the command-line client, the Java API, and the Sarama API call it `ResetOffset`?

Thanks for clarifying, I responded quickly without doing a good job of reviewing your comment thoroughly (need to slow down ; ).  There's no good reason for the naming difference, you are right that they should align and I will change the terminology (ResetOffset, etc.) - thanks!  

> 
> Also, If we create a CRD that supports multiple commands, I see an increased complexity in:
> 
> * granting access to individual commands rather than all of them.
> * API or command versioning
 
I am open to having the Command CRD be very specific `KafkaResetOffsetCommand` (or similar) as you suggest - I thought I heard the opposite in previous conversations.  The reasons you provided are good ones, but there are pros/cons for each approach.  Anyone else have a strong opinion?
 
> > I have some research to do in that regard, but assuming it is feasible then we are able to eliminate the need for the control-protocol, which is still POC level and might not be ready in time for this work. I'm interested in everyone's opinion as to the acceptability of that versus what we'd been discussing related to the control-protocol? Thanks again everyone - closing in on it ; )
> 
> I haven't followed any conversation, so 🤷‍♂️, but this was my assumption, it's a control plane only action.

Yeah, my initial assumption was that this had to be handled in the Dispatcher logic where all the ConsumerGroup stuff is today.  The only reasons to do it there would be to keep the ConsumerGroup logic together, and to be able to cleanly stop the ConsumerGroups before moving the offsets.  If those concerns can be ignored, then the Controller can just move the offsets without all the `Controller --> Dispatcher` communication which is simpler.

@slinkydeveloper - This Controller only approach was only being considered to alleviate the time-crunch of getting the `control-protocol` in place soon enough for this replay work.  If we can make the schedule / timelines work, then I am fine with the original plan.  It is a bit more complex, but does have benefits.  I will write the Feature Track document with both options and we can discuss schedules when we review that?  Again, I don't mean to impose my internal deadlines on everyone else, but I do have commitments to meet and so was considering various options.

> The race condition is referring to I guess is related to the fact that they need to support the `--shift-by n` option, which is out of scope for us.
> 
> The "is active" check is client-side https://github.com/apache/kafka/blob/617ee003223873da2c1c7f383d3416684c2d84c8/core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala#L435-L453, so I guess we can do the reset directly to a specific timestamp using Sarama on the control plane without pausing consumers.

Ah, ok that makes sense - thanks for providing the background info.  I agree we don't need the `--shift-by` option and can get away without pausing the consumers.  It makes sense to stop them gracefully (internally) if we can, but is probably not required.

**Comment by slinkydeveloper** [comment]
> This Controller only approach was only being considered to alleviate the time-crunch of getting the control-protocol in place soon enough for this replay work. If we can make the schedule / timelines work, then I am fine with the original plan. It is a bit more complex, but does have benefits.
> Again, I don't mean to impose my internal deadlines on everyone else, but I do have commitments to meet and so was considering various options.

Is there something of the control protocol that is not in place/doesn't fit this use case? Can you be more specific on what doesn't fit?

I personally disagree with the idea of merging code, just for the sake of following whatever schedule, which brings in immediately a lot of tech debt that has to be solved at some point by somebody... In other words, I disagree with taking shortcuts... If control-protocol is the tool that fits the problem, we should use it and invest all the time we need to stabilize it and make it fit the use case.

**Comment by matzew** [comment]
>  If control-protocol is the tool that fits the problem, we should use it and invest all the time we need to stabilize it and make it fit the use case. If it doesn't fit, then we don't use it.

that seems reasonable to me as well.

**Comment by travis-minke-sap** [comment]
> Is there something of the control protocol that is not in place/doesn't fit this use case? Can you be more specific on what doesn't fit?

I never indicated that the control-protocol had a deficiency or didn't fit the use case.  I was only stating that there are other approaches that might be more expedient to complete.  I was basing my assumption that the control-protocol **might** not be ready in time on your proposed ordering of work - the last time we spoke you said the control-protocol was still under construction, and you wanted to roll it out into the KafkaChannels as part of building the Pause / Resume feature and you couldn't commit to a timeline for that?  I know Matthias was going to meet with you to see if that was still true?  I didn't hear back definitively what the outcome was - only that I should write a Feature Track document (which I have started). 
 
> I personally disagree with the idea of merging code, just for the sake of following whatever schedule, which brings in immediately a lot of tech debt that has to be solved at some point by somebody... In other words, I disagree with taking shortcuts... If control-protocol is the tool that fits the problem, we should use it and invest all the time we need to stabilize it and make it fit the use case. If it doesn't fit, then we don't use it.

Yeah, in an ideal world where there are no deadlines it's fine to do things in the correct order, but... I don't live in that world ; ) and have internal commitments to meet.  I also realize it's not in any way fair of me to expect the larger project to align with my internal deadlines.  I can contribute time to integrating the control-protocol as long as I have some support and the schedules work.

I'm really not trying to change direction on you - I was just reacting to Pier's Controller-only approach by brainstorming alternatives.  I am fine with pursuing the control-protocol approach as long as have a path to completion in the next few months.  Sorry for the headache - I appreciate everyone's time on this!

**Comment by travis-minke-sap** [comment]
This was an initial first step in the process of the design/implementation of the Replay feature and is no longer needed beyond historical information.  Work on ResetOffset (Replay) is progressing nicely in #588 so I will close this Issue ; )
